# Основы Python

## Командная оболочка Python

В этой главе рассматриваются основы синтаксиса, структуры и типы данных языка программирования Python. В нескольких первых разделах предполагается написание не более чем нескольких строк кода Python, которые могут быть выполнены с использованием командной оболочки (shell) языка Python. Это интерактивная рабочая среда: пользователь вводит инструкции на языке Python, которые выполняются немедленно сразу после нажатия клавиши Enter.

Действия, необходимые для получения доступа к встроенной командной оболочке языка Python, отличаются в различных операционных системах. Для запуска командной оболочки из командной строки сначала необходимо открыть окно терминала (консоли), воспользовавшись затем ввести команду python.

Для выхода из командной оболочки Python выполните команду exit(). После запуска командной оболочки Python вы увидите приветственное сообщение (которое может меняться в зависимости от используемой операционной системы и версии Python). 

Три правые угловые скобки (>>>) – это приглашение, или промпт (prompt), после этих символов приглашения вы будете вводить команды языка Python. В данной книге используется версия Python 3, поэтому вы должны проверить номер версии в первой строке – Python 3.X.Y: здесь не особенно важны значения X и Y, представляющие минорный номер версии.

Многие дистрибутивные пакеты Python содержат немного более продвинутую командную оболочку, которая называется IDLE, с дополнительными функциями завершения по нажатии клавиши Tab и подсветкой синтаксиса (syntax highlighting) (ключевые слова выделяются различными цветами, когда вы вводите их). Но мы не будем рассматривать это приложение, а предпочтем ему более новую и более усовершенствованную рабочую программную среду IPython.

Кроме того, для многих вариантов установки существует возможность (особенно в Windows) запуска командной оболочки Python непосредственно из приложения, установленного в процессе установки самого интерпретатора Python. Некоторые варианты установки даже добавляют ярлык со значком на рабочий стол, так что можно открыть командную оболочку Python щелчком (или двойным щелчком) по этому ярлыку.

## Числа, переменные, операции сравнения и логические операции

## Типы числовых значений

Одним из самых простых объектов языка Python являются числовые значения, для которых определены три типа: целые числа (тип: int), числа с плавающей точкой (тип: float) и комплексные числа (тип: complex).

### Целые числа

Целые числа (integers) – это математические целые числа, такие как 1, 8, −72, 3 847 298 893 721 407. В версии Python 3 нет ограничений по их величине (кроме ограничения по доступности оперативной памяти компьютера). Арифметика целых чисел является точной.

Для удобства разрешается разделять группы цифр (разрядов) целого числа символом подчеркивания «_». Например, запись 299_792_458 интерпретируется как целое число 299 792 458.

## Числа с плавающей точкой

Числа с плавающей точкой (floating-point numbers) являются представлением действительных чисел, таких как 1.2, −0.36 и 167263×10−7. Вообще говоря, не существует точных значений действительных чисел в представлении Python, но эти числа хранятся в бинарном (двоичном) виде с определенной точностью (в большинстве систем точность составляет 15–16 разрядов), как описано в разделе 10.1. Например, дробь 4/3 хранится как двоичное равнозначное представление 1.33333333333333325931846502…, которое приблизительно (но не точно) равно бесконечно повторяющемуся десятичному представлению дроби 4/3 = 1.3333…. Более того, даже числа, для которых существует точное десятичное представление, могут не иметь точного бинарного представления: например, дробь 1/10 представлена бинарным числом, равным значению 0.10000000000000000555111512…. Из-за такой ограниченной точности арифметика чисел с плавающей точкой не является абсолютно точной, но при аккуратном отношении она «достаточно точна» для большинства научных приложений.

Любое отдельное число, содержащее символ точки (.), рассматривается в Python как представление числа с плавающей точкой. Также поддерживается научный формат записи чисел с использованием буквы e или E для отделения

Это соответствует реализации по международному стандарту представления чисел с плавающей точкой двойной точности IEEE-754.

## Числа, переменные, операции сравнения и логические операции

В значимой части числа (мантиссы) от показателя степени: например, 1.67263e-7 представляет число 1 67263 × 10−7. Как и для целых чисел, группы разрядов можно разделять символами подчеркивания. Например, 1.602_176_634e-34.

## Комплексные числа

Комплексные числа, такие как 4+3j, состоят из действительной и мнимой частей (в Python мнимая часть обозначается буквой j), каждая из которых сама по себе является числом с плавающей точкой (даже если записана без символа точки). Таким образом, арифметика комплексных чисел не является точной, но обеспечивает такую же конечную ограниченную точность, что и числа с плавающей точкой (float).

Комплексное число может быть определено «сложением» действительного числа с мнимым (обозначенным суффиксом j): 2.3 + 1.2j – или посредством отдельной передачи действительной и мнимой частей при вызове complex, на-пример complex(2.3, 1.2).

**Пример**. Ввод числа после приглашения командной оболочки Python просто возвращает это же число

```python
>>> 5
5
>>> 5.
5.0
>>> 0.10
0.1
>>> 0.0001
0.0001
>>> 0.0000999
9.99e-05
```

Следует отметить, что интерпретатор Python выводит числа стандартным способом. Например: Внутреннее представление числа 0.1, описанное выше, округляется до 0.1, т. е. до самого короткого числового значения в этом представлении.

Числа, меньшие по величине, чем 0.0001, выводятся в научном формате.

Число любого типа может быть создано из числа другого типа с помощью соответствующего конструктора (constructior):


```python
>>> float(5)
5.0
>>> int(5.2)
5
>>> int(5.9)
3j
5

>>> complex(3.)
(3+0j)
>>> complex(0., 3.)

```

Обратите внимание: положительное число с плавающей точкой округляется с недостатком (в меньшую сторону) во время преобразования его в целое число. Более общее правило: метод int округляет в сторону нуля:

```int(-1.4)``` дает результат -1.

Создание объекта типа complex из объекта типа float генерирует комплексное число с мнимой частью, равной нулю.

Для генерации абсолютно мнимого числа необходимо явно передать два числа в метод ```complex```, при этом первая, действительная часть должна быть равна нулю.

# Использование командной оболочки Python в качестве калькулятора

## Простые арифметические действия

Описанными в предыдущем разделе тремя основными типами чисел можно использовать командную оболочку Python как простой калькулятор, применяя операторы, описанные в табл. 1. Это бинарные операторы, поскольку они работают с двумя числами (операндами) для создания третьего числа (например, при вычислении ```2**3``` получается результат ```8```).

### Таблица 1. Основные арифметические операторы языка Python

|      |                             |
|------|:-----------------------------| 
|+     |Сложение                     |
|-     |Вычитание                    |
|*     |Умножение                    |
|/     |Деление с плавающей точкой   |
|//    |Целочисленное деление        |
|%     |Деление по модулю            |
|**    |Возведение в степень         |

В версии Python 3 существует два типа операции деления: деление чисел плавающей точкой (/) всегда возвращает как результат число с плавающей точкой (или комплексное число), даже если применяется к целым числам.

Целочисленное деление (//) всегда округляет результат с недостатком (в меньшую сторону), т. е. к ближайшему меньшему целому числу (floor division). Типом результата является int, только если оба операнда имеют тип int, иначе возвращается значение типа float. Ниже приведены примеры, помогающие понять эти правила.

Обычное («истинное») деление с плавающей точкой с использованием оператора (/):

In [8]:
2.7/2

1.35

In [9]:
9 / 2

4.5

In [10]:
8 / 4

2.0

## Числа, переменные, операции сравнения и логические операции

Последняя операция деления возвращает значения типа float даже притом, что оба операнда имеют тип int.

Целочисленное деление с использованием оператора (//):

In [12]:
8//4

2

In [13]:
9//2

4

In [14]:
2.7 // 2

1.0

Обратите внимание: оператор // может применять целочисленную арифметику (с округлением в меньшую сторону) к числам с плавающей точкой.

Оператор деления по модулю (%) возвращает остаток от целочисленного деления:

In [15]:
9 % 2

1

In [16]:
4.5%3

1.5

в этом случае возвращается значение типа int, только если оба операнда имеют тип int.

## Приоритет операторов

Арифметические операции могут быть объединены в последовательность, из-за чего естественным образом возникает вопрос приоритета выполнения операций: например, при вычислении выражения 2 + 4 * 3 должно получиться 14 (в результате 2 + 12) или 18 (в результате 6 * 3)? Из табл. 2 следует, что результат должен быть равен 14, так как операция умножения имеет более высокий приоритет, чем операция сложения, поэтому должна выполняться первой. Эти правила приоритета изменяются при использовании круглых скобок, например (2 + 4) * 3 = 18.

### Таблица 2. Приоритеты арифметических операторов языка Python

|    |    |
|:----|:----|
|**|Самый высокий приоритет|
|*, /, //, % ||
|+, − |Самый низкий приоритет|

Операторы с одинаковым приоритетом выполняются слева направо, за исключением операции возведения в степень (**), которая выполняется справа налево (т. е. «сверху вниз» при записи в обычной математической форме показателей степени над строкой). Например:

In [19]:
6 / 2 / 4 # равнозначно 3 / 4

0.75

In [20]:
6 / (2 / 4) # равнозначно 6 / 0.5

12.0

In [21]:
 2**2**3 # равнозначно 2**(2**3) == 2**8


256

In [22]:
 (2**2)**3 # равнозначно 4**3

64

В показанных здесь примерах символ «решетка» (#) обозначает начало комментария, который игнорируется интерпретатором. Комментарии иногда будут использоваться для более подробного объяснения конкретной инструкции,
но при вводе кода для выполнения комментарии вводить не обязательно.

## Методы и атрибуты числовых значений

Числовые значения в языке Python являются объектами (objects) (в действительности в Python все является объектами) и имеют определенные атрибуты (attributes), доступные при использовании формата «точка» (dot notation): <объект>.<атрибут> (такое использование точки не имеет ничего общего с десятичной точкой в числах с плавающей точкой). Некоторые атрибуты являются простыми значениями, например объекты комплексных чисел имеют атрибуты ```real``` и ```imag```, соответствующие действительной и мнимой частям (с плавающей точкой) комплексного числа:

In [23]:
(4 + 5j).real

4.0

In [24]:
(4 + 5j).imag

5.0

Другими атрибутами являются методы (methods): вызываемые функции, которые определенным образом работают со своим объектом. Например, для комплексных чисел существует метод ```conjugate```, который возвращает сопряженное комплексное число:

In [26]:
(4 + 5j).conjugate()

(4-5j)

Здесь пара пустых круглых скобок обозначает свойство вызываемости метода, т. е. выполнения вычисления сопряженного комплексного числа для числа ```4 + 5j```. Если скобки не указаны, например ```(4 + 5j).conjugate```, то мы получаем ссылку на сам метод (без его вызова) – ведь этот метод тоже является объектом.

В действительности целые числа и числа с плавающей точкой не обладают слишком многими атрибутами, которые имеет смысл использовать показанным выше способом, но если вам интересно узнать, сколько битов занимает целое число в оперативной памяти, то можно воспользоваться методом ```bit_length```. Например:

In [27]:
(3847298893721407). bit_length()

52

Следует отметить, что Python выделяет столько памяти, сколько необходимо для точного представления целого числа.

В этой книге термины метод (method) и функция (function) используются как взаимозаменяемые. В языке Python все является объектами, поэтому различие между этими терминами не столь существенно, как в некоторых других языках программирования.


## Математические функции

Две математические функции из множества, предоставляемые «по умолчанию» как встроенные (built-in) функции, – ```abs``` и ```round```.

Функция ```abs``` возвращает абсолютное значение числа:

In [28]:
abs(-5.2)

5.2

In [29]:
abs(-2)

2

In [30]:
abs(3 + 4j)

5.0

Это пример полиморфизма (polymorphism): одна и та же функция ```abs``` выполняет различные операции с различными объектами. Если в функцию передано действительное число ```x```, то возвращается ```|x|```, неотрицательная величина этого числа без учета знака. Если передается комплексное число ```z = x + iy```,то возвращается его модуль ```z = sqrt(x2 + y2)```.

Функция ```round``` (с одним аргументом) округляет число с плавающей точкой до ближайшего целого числа, используя для этого метод округления банкира (Banker’s rounding)

In [31]:
round(-9.62)

-10

In [32]:
round(7.5)

8

In [33]:
round(4.5)

4

Можно также задать количество точных разрядов после десятичной точки как второй аргумент, передаваемый в функцию round():

In [35]:
round(3.141592653589793 , 3)

3.142

In [36]:
round(96485.33289, -2)

96500.0

При округлении по методу банкира числа «с 0.5» всегда округляются до ближайшего четного целого числа.

Python представляет собой в высшей степени модульный язык: функциональность доступна в пакетах и модулях, которые при необходимости импортируются, но по умолчанию не загружаются: это позволяет экономить память, требуемую для работы Python-программы, сохраняя ее размер на минимальном уровне, и улучшать производительность. Например, многие полезные математические функции расположены в модуле ```math```, который импортируется следующей командой:

Модуль ```math``` включены операции для работы с числами с плавающей точкой и целыми числами (для функций, работающих с комплексными числами, существует другой модуль ```cmath```). Функции вызываются с передачей одного числа (иногда нескольких чисел) внутри круглых скобок (числа принимаются как аргументы вызываемой функции). Например:

In [37]:
import math
math.exp(-1.5)

0.22313016014842982

In [38]:
math.cos(0)

1.0

In [39]:
math.sqrt(16)

4.0

Полный список математических функций, предоставляемых модулем ```math```, можно найти в онлайновой документации. Некоторые наиболее часто используемые математические функции перечислены в табл. 3.

Таблица 3. Некоторые функции, предоставляемые модулем ```math```. Предполагается, что аргументы, соответствующие значениям углов, задаются в радианах

|          |               |
|:----------|:---------------|
|math.sqrt(x)     |√x    |
|math.exp(x)      |ex|
|math.log(x)      |ln x|
|math.log(x, b)                  |logb x          |
|math.log10(x)                  |log10 x          |
|math.sin(x)                |sin(x          |
|math.cos(x)                 |cos(x          |
|math.tan(x)                 |tan(x)          |
|math.asin(x)                 |arcsin(x)          |
|math.acos(x)                 |arccos(x)          |
|math.atan(x)                  |arctan(x)          |
|math.sinh(x)                 |sinh(x)          |
|math.cosh(x)                 |cosh(x)          |
|math.cosh(x)                 |cosh(x)          |
|math.tanh(x)                 |tanh(x          |
|math.asinh(x)                 |arsinh(x          |
|math.acosh(x)                 |arcosh(x)          |
|math.atanh(x)                 |artanh(x)          |
|math.hypot(x, y)                 |Евклидово нормальное расстояние (гипотенуза) √(x2+ y2)          |
|math.factorial(x)                 |Факториал x!          |
|math.erf(x)                 |Функция ошибок по x          |
|math.gamma(x)                 |Гамма-функция по x, Γ(x)          |
|math.degrees(x)                  |Преобразование x из радианов в градусы          |
|math.radians(x)                 |Преобразование x из градусов в радианы          |
|math.isclose(a, b)                  |Проверка равенства a и b в пределах некоторого допустимого отклонения          |





Кроме того, модуль math предоставляет два весьма полезных нефункциональных атрибута: ```math.pi``` и ```math.e``` – это значения математических констант и ```e``` (основание натуральных логарифмов).

Литература: https://docs.python.org/3/library/math.html.

Существует возможность импорта модуля ```math``` с помощью команды ```from math import *```, после чего становится доступным прямой доступ ко всем его функциям:

In [42]:
from math import *
cos(pi)

-1.0

Это может оказаться удобным для работы в командной оболочке Python, но не рекомендуется в программах на языке Python. Возникает опасность конфликтов имен (особенно если таким способом импортируются функции из многих модулей), и становится трудно узнать, из какого модуля импортирована та или иная функция. Импорт командой ```import math``` сохраняет связь функций пространством имен (namespace) соответствующего модуля, поэтому, даже несмотря на то что ввод ```math.cos``` требует больше нажатий на клавиши, такой способ делает исходный код более удобным для понимания и сопровождения.

**Пример 2**. Вполне естественно, что математические функции можно объединять в одном выражении

In [43]:
import math
math.sin(math.pi/2)

1.0

In [44]:
math.degrees(math.acos(math.sqrt(3)/2))

30.000000000000004

Обратите внимание на ограниченную (конечную) точность последнего выражения: точный результат ```arccos(√3/2) = 30°```.

Тем фактом, что функция ```int``` округляет в меньшую сторону при преобразовании положительного числа с плавающей точкой в целое число, можно воспользоваться для определения количества разрядов положительного целого числа:

In [45]:
int(math.log10(9999)) + 1

4

In [46]:
int(math.log10(10000)) + 1

5

# Переменные

## Что такое переменная

При создании объекта, например типа ```float```, в программе на Python или при использовании командной оболочки Python для этого объекта выделяется память: в компьютерной архитектуре место расположения данного фрагмента памяти называется адресом памяти (address). Действительное значение адреса памяти объекта не очень полезно (и не очень удобно) в Python, но если вы захотите, то можете узнать адрес, вызвав встроенный метод ```id```:

In [48]:
id(20.1)

2906346245776

Это число является ссылкой на конкретное место в памяти, соответствующее фрагменту памяти, выделенному для хранения объекта типа ```float``` со значением 20.1.

Кроме самых простых вариантов использования, существует необходимость в хранении объектов, участвующих в вычислении или в выполнении алгоритма, к которым можно было бы обращаться по некоторому более удобному и осмысленному имени (а не по числовому адресу в памяти). Для этого существуют переменные (variables). Имя переменной может быть присвоено любому объекту («связано» с любым объектом) и использоваться для идентификации этого объекта в дальнейших вычислениях. Например:

In [51]:
a = 3
b = -0.5
a * b

-1.5

В этом фрагменте кода создается объект типа ```int``` со значением ```3```, и ему присваивается имя ```a```. Затем создается объект типа ```float``` со значением ```-0.5```, и ему присваивается имя ```b```. Выполняется вычисление ```a * b```: значения ```a``` и ```b``` умножаются друг на друга, и возвращается результат. Этот результат не присваивается никакой переменной, сразу после вывода на экран он отбрасывается. Таким образом, для хранения результата память не требуется, а для значения ```-1.5``` типа ```float``` выделяется память только на то короткое время, которое необходимо для предъявления результата пользователю, затем память освобождается, значение теряется. Если результат необходим для дальнейших вычислений, то вы должны присвоить его другой переменной:


In [52]:
c = a * b
c

-1.5

Обратите внимание: не требуется предварительное объявление (declare) переменных до присваивания им значений (т. е. для сообщения Python о том, что имя переменной a является ссылкой на целое число, имя переменной ```b``` – ссылкой на число с плавающей точкой и т. д.), как это необходимо делать в некоторых языках программирования. Python – язык с динамической типизацией (dynamically typed), поэтому необходимый тип объекта логически выводится из его определения: при отсутствии десятичной точки число 3 определяется как ```int``` . Число -0.5 выглядит как число с плавающей точкой, и Python определяет ```b``` как ```float```.

## Имена переменных

Ниже приводится несколько правил, определяющих формирование допустимых («правильных») имен переменных: в именах переменных учитывается регистр символов (букв): ```a``` и ```A``` – это разные имена переменных; имена переменных могут содержать любую букву, символ подчеркивания (_) и любую цифру (0–9 но не должны начинаться с цифры; имя переменной не должно совпадать с одним из зарезервированных ключевых слов, приведенных в табл. 4.

Встроенные имена констант ```True```, ```False``` и ```None``` запрещено использовать как имена переменных.

### Таблица 4. Зарезервированные ключевые слова языка Python 3

|          |           |            |
|----------|-----------|------------|
|and         |as           |assert            |
|async          |await           |break            |
|class          |continue           |def            |
|del          |elif           |else            |
|except          |finally           |for            |
|from          |global           |if            |
|import          |in           |is            |lambda
|nonlocal          |not           |or            |pass
|raise          |return           |try            |while
|yield          |with           |False            |True
|None          |           |            |


Большинство зарезервированных ключевых слов вряд ли подходят для имен переменных, за исключением разве что слова ```lambda```. Программисты на Python часто используют имя ```lambda``` при необходимости. Текстовый редактор с расширенными функциями подсвечивает ключевые слова при вводе исходного кода программы, поэтому путаница с именами возникает редко. Существует возможность присваивания переменной имени, совпадающего именем встроенной функции (например, ```abs``` или ```round```), но после такого присваивания встроенная функция становится недоступной, поэтому присваивания подобных имен лучше избегать – к счастью, большинство встроенных функций имеют имена, которые вряд ли будут выбираться в реальной практике.

В дополнение к перечисленным выше правилам ниже приводятся некоторые соглашения по стилю, определяющие общепринятые практические принципы именования переменных:

- имена переменных должны быть осмысленными (```area``` лучше, чем ```a```)…
- …но не слишком длинными (```the_area_of_the_triangle``` – это слишком громоздкое имя);
- в общем случае лучше не использовать ```I``` (буква ```i``` в верхнем регистре), ```l``` (буква ```L``` в нижнем регистре) и букву ```O``` в верхнем регистре, так как они очень похожи на цифры 1 и 0.

- имена переменных ```i```, ```j``` и ```k```, как правило, используются для целочисленных счетчиков;

- рекомендуется использовать имена с буквами нижнего регистра с разде-лением слов символами подчеркивания вместо стиля именования «Cam-elCase»: например, ```mean_height```, а не ```MeanHeight```.


Полный список имен встроенных функций см. в документации: https://docs.python.org/3/library/functions.html. 


Эти и многие другие правила и соглашения определены в руководстве по стилю под названием PEP8, которое представляет собой одну из частей доку-ментации Python.

Нарушение этих правил хорошего стиля не приведет к нарушению правильной работы программы, но такую программу, возможно, труднее будет сопровождать и отлаживать – может быть, вы помогаете самому себе, соблюдая правила стиля.

**Пример 3**. Формула Герона для вычисления площади ```A``` треугольника со сторонами ```a```, ```b```, ```c```:

$ A = \sqrt {s(s-a)(s-b)(s-c)}  $ , где $ s = \frac{1}{2}(a+b+c) $


In [58]:
a = 4.503
b = 2.377
c = 3.902
s = (a+b+c)/2
area = math.sqrt(s*(s-a)*(s-b)*(s-c))
area

4.63511081571606

**Пример 4**. Тип данных и адрес памяти объекта по имени соответствующей переменной можно получить с помощью встроенных функций ```type``` и ```id```:

In [59]:
type(a)

float

In [60]:
id(area)

2906348016048

# Операции сравнения и логические операции

## Операторы

Основные операторы, используемые в Python для сравнения объектов (например, чисел), перечислены в табл. 5.

### Таблица 5. Операторы сравнения Python

|    |    |
|:----|:----|
|==|Равно|
|!= | Не равно |
| < | Меньше |
| > | Больше |
| >= | Больше равно |
| <= | Меньше равно |

Результатом сравнения является логический (boolean) объект (или объект типа ```bool```), который может содержать одно и только одно из двух возможных значений: ```True``` или ```False```. Это встроенные постоянные ключевые слова, которым не могут быть присвоены другие значения. Например:

In [61]:
7==8

False

In [62]:
4 >= 3.14

True

Python способен, если это возможно без двусмысленности, сравнивать объекты различных типов: целое число 4 приводится к типу ```float``` для сравнения с числом 3.14.
Обратите особое внимание на различие между оператором ```==``` и оператором ```=```. Один знак равенства – это присваивание (assignment), операция, которая не возвращает значение: инструкция ```a = 7``` присваивает имя переменной ```a``` целочисленному объекту 7, и это все. Выражение ```a == 7``` – это проверка условия: возвращается значение ```True``` или ```False```, в зависимости от значения переменной ```a```.

Особое внимание требуется при сравнении чисел с плавающей точкой на равенство. Поскольку числа с плавающей точкой хранятся с некоторой неточностью, вычисления с ними часто приводят к потере точности, и можно получить неожиданные результаты по неосторожности. Например:


In [63]:
a = 0.01
b = 0.1**2
a == b

False

В этом примере 0.01 не может быть представлено точно как число с плавающей точкой, а хранится (в моей системе) как бинарное число, равное зна-чению 0.010000000000000000208, а с другой стороны, результат возведения в квадрат представления значения 0.1 в виде числа с плавающей точкой равен 0.01000000000000000194, и эти два числа не равны.

В некоторых языках, например в C, операция присваивания возвращает значение, которое присваивается переменной, и это может приводить к некоторым опасным и трудно обнаруживаемым ошибкам, когда оператор = используется ошибочно, как оператор сравнения.
 В версии Python 3.5 в библиотеке math появилась функция ```isclose```, которая проверяет равенство двух чисел с плавающей точкой с учетом некоторого абсолютного или относительного интервала допустимого отклонения:

In [64]:
math.isclose(0.1**2, 0.01)

True

Величина относительного допустимого отклонения может быть установлена помощью аргумента ```rel_tol```, по умолчанию равного 1.e-9: это максимально допустимая разность между двумя числами относительно абсолютного значения большего из этих чисел. Например, для проверки равенства ```a``` и ```b``` в пределах 5 % от большего из них:

In [65]:
a = 9.5
b = 10
math.isclose(a, b, rel_tol=0.05)

True

При этом типе относительного сравнения могут возникать проблемы, если одно из чисел равно нулю. В этом случае может оказаться полезной проверка абсолютного допустимого отклонения, значение которого передается в аргументе abs_tol (по умолчанию 0):

In [66]:
math.isclose(0, 1.e-12) # сравнение по относительному допустимому отклонению дает неверный результат, если rel_tol < 1


False

In [67]:
math.isclose(0, 1.e-12, abs_tol=1.e-10)

True

# Логические операторы

Операции сравнения можно изменять и объединять с помощью ключевых слов, обозначающих логические операторы: ```and```, ```not``` и ```or```. См. табл. 6, 7 и 8.

### Таблица 6. Таблица истинности для оператора not

|P     |not P |
|:-----|:-----|
|True  |False |
|False |True  |


### Таблица 7. Таблица истинности для оператора and

|P     |Q      |P and Q |
|:-----|:------|:-------|
|True  |True   |True    |
|False |True   |False   |
|True  |False  |False   |
|False |False  |False   |

### Таблица 8. Таблица истинности для оператора or

|P     |Q      |P or Q  |
|:-----|:------|:-------|
|True  |True   |True    |
|False |True   |True    |
|True  |False  |True    |
|False |False  |False   |

Например:

In [70]:
7.0 > 4 and -1 >= 0# равнозначно True and False

False

In [72]:
5 < 4 or 1 != 2 # равнозначно False or True


True

В составных выражениях, подобных приведенным выше, первыми выполняются операторы сравнения, а затем логические операторы в порядке их приоритетности: ```not```, ```and```, ```or```. Приоритет можно изменить с помощью круглых скобок, как для арифметических выражений. Например:

In [73]:
not 7.5 < 0.9 or 4 == 4

True

In [74]:
not (7.5 < 0.9 or 4 == 4)

False

Таблицы 6, 7 и 8 – это таблицы истинности для логических операторов. Следует отметить, что, как и в большинстве языков программирования, в Python оператор ```or``` представляет собой вариант «включающее или» (inclusive or), при котором A or B равно ```True```, если A, или B, или и A, и B истинны, в отличие от оператора «исключающее или» (exclusive or) (A xor B равно ```True```, только если либо A, либо B истинно, но не оба операнда одновременно).

**Замечание**: в Python нет встроенного оператора ```xor```, но его можно импортировать как функцию с помощью инструкции ```from operator import xor```. Вызов этой функции ```xor(a, b)``` возвращает ```True``` или ```False```.
 


## Логические равенства и условное присваивание

В выражении логической проверки не всегда необходимо выполнять явное сравнение для получения логического значения: Python попытается преобразовать объект в тип ```bool```, если это необходимо. Для числовых объектов 0 преобразуется в ```False```, а любое ненулевое значение – в ```True```:

In [75]:
a = 0

In [76]:
a or 4 < 3 # равнозначно: False or 4 < 3

False

In [77]:
not a + 1 # равнозначно: not True


False

В этом примере операция сложения имеет более высокий приоритет, чем логический оператор ```not```, поэтому сначала вычисляется ```a+1``` и дает результат 1. Это соответствует логическому значению ```True```, следовательно, все выражение равнозначно выражению ```not True```. Для явного преобразования любого объекта в логический объект используется конструктор ```bool```:

In [78]:
 a = 0

In [79]:
a - 2 or a

-2

In [80]:
4 > 3 and a - 2

-2

In [81]:
4 > 3 and a

0

Логические выражения вычисляются слева направо, при этом предполагается, что вычисления ```and``` и ```or``` выполняются по укороченной схеме (shortcir-cuited): второе выражение вычисляется, только если необходимо определить истинное значение всего выражения в целом. Приведенные выше три примера можно проанализировать следующим образом:

- В  первом примере сначала вычисляется a-2: результат равен −2, т.  е. равнозначен значению True, поэтому условие or выполнено, и операнд, вычисленный как True, возвращается немедленно: −2.
- Проверка условия  4 > 3 дает результат True, поэтому должно быть вычислено второе выражение, чтобы определить истинность условия and.Выражение a-2 равно −2, что также равнозначно True, следовательно, условие and выполнено, и возвращается −2 (как результат выражения, вычисленного самым последним).
- В последнем случае a равно 0, что равнозначно False, поэтому результатом вычисления условия and становится False,и возвращается значение 0.

## None – специальное значение в языке Python

В Python определено отдельное специальное значение None с особенным типом ```NoneType```. Оно используется для представления отсутствия определенного значения, например в случаях, когда не существует возможного значения,
и других подобных случаях. Это особенно удобно, когда нужно избежать введения произвольных значений по умолчанию (таких как 0, -1 или -99) для испорченных или отсутствующих данных.
В выражениях логического сравнения ```None``` вычисляется как ```False```, но для проверки равенства значения переменной x  значению ```None``` необходимо использовать конструкцию

```if x is None```
и
```if x is not None```
вместо сокращенных вариантов ```if``` x и ```if not x```

**Замечание**: Обратите внимание: ```not x``` также вычисляется как ```True```, если x является любым из
следующих объектов: ```0```, ```False``` или пустая структура данных, например пустой список.


**Пример 5**. Общеизвестный прием в Python – присваивание переменной значения, возвращаемого логическим выражением:

In [83]:
a = 0
b = a or -1
b

-1

Это буквально означает (предполагается, что a должно быть целым числом):
«установить b равным значению a, если не выполнено условие a == 0, а в случае
выполнения этого условия установить значение b равным −1».

## Неизменяемость и идентичность

Объекты, рассматриваемые до настоящего момента, такие как целые числа и логические значения, являются неизменяемыми (immutable).Неизменяемые объекты никогда не изменяются после создания, хотя имя переменной может быть переназначено для ссылки на другой объект, отличающийся от объекта, которому было изначально присвоено это имя. Например, рассмотрим следующие присваивания:

In [85]:
a = 8
b = a

В первой строке создается целочисленный объект со значением 8 в памяти, и ему присваивается имя a. Во второй строке тому же объекту присваивается имя b. Это можно увидеть, проверив адрес памяти объекта по каждому имени:

In [86]:
id(a)

2906229402128

In [87]:
id(b)


2906229402128

Таким образом, a и b – ссылки на один и тот же целочисленный объект. Теперь предположим, что имя a переназначается для нового числового объекта:

In [88]:
a = 3.14

In [89]:
 a

3.14

In [90]:
b

8

In [91]:
id(a)

2906348018544

In [92]:
id(b)

2906229402128

Обратите внимание: значение b не изменилось –эта переменная продолжает ссылаться на исходное значение 8. Переменная a теперь ссылается на новый объект типа float со значением 3.14,расположенный по новому адресу памяти. Именно это подразумевается под неизменяемостью: это не «переменная», которая не может изменяться, а сам объект является неизменяемым

![image](image.png)

Рис. 2.1. (а) Две переменные ссылаются на одно и то же целое число; (б) после переназначения для переменной a другого значения

Более удобный способ установления того факта, что две переменные ссылаются на один и тот же объект, – использование оператора ```is```, который определяет идентичность (identity) объектов:

In [1]:
a = 2432
b = a
a is b

True

In [2]:
c = 2432
c is a

False

In [3]:
c == a

True

Здесь присваивание ```c = 2432``` создает абсолютно новый объект, поэтому вычисление выражения ```c is a``` дает результат ```False```, даже притом, что ```a``` и ```c``` содержат одинаковое значение. Таким образом, эти две переменные ссылаются на
различные объекты с одинаковыми значениями. Часто необходимо изменить значение переменной каким-либо способом,
например:

In [4]:
a = 800
a = a + 1
a

801

Целые числа 800 и 801 неизменяемы: строка ```a = a + 1``` создает новый целочисленный объект со значением 801 (правая сторона выражения вычисляется в первую очередь) и присваивает его переменной с именем ```a```  (старое значение 800 «забывается», если только на это значение не ссылается какая-либо другая переменная). Таким образом, ```a``` указывает на разные адреса памяти до
и после выполнения этой инструкции. Подобное переназначение переменной на результат выполнения арифметической операции выполняется настолько часто, что была введена удобная укороченная форма ее записи: комбинированное присваивание (augmented
assignment) ```a += 5``` равнозначно присваиванию ```a = a + 5```. Точно так же работают
операторы ```-=, *=, /=, //=, %=```. Но в Python не поддерживаются операции инкремента и декремента в C-стиле, такие как ```a++``` для операции ```a += 1```.

**Пример 6**. Python предоставляет оператор ```is not```: более естественно написать
```c is not a```, чем ```not c is a```

In [5]:
a = 8
b = a
b is a


True

In [6]:
b /= 2
b is not a

True

**Пример 7**. С учетом приведенного выше описания может показаться странным
следующий факт:

In [7]:
a = 256
b = 256
a is b

True

Причина в том, что Python сохраняет кеш (cache) общего пользования для небольших целочисленных объектов (в моей системе это числа от -5 до 256). Для повышения производительности присваивание ```a = 256``` связывает имя переменной a с существующим целочисленным объектом без необходимости выделения новой памяти для него. Поскольку такое же присваивание выполняется и для переменной b, эти две переменные в данном конкретном случае в  действительности указывают на один и тот же объект. Противоположный пример:

In [8]:
a = 257
b = 257
a is b

False

# Упражнения

## Вопросы

### В 1. Определить результаты вычислений следующих выражений и проверить их, используя командную оболочку Python.

- а)	 2.7 / 2
- б) 2 / 4 - 1
- в) 2 // 4 - 1
- г) (2 + 5) % 3
- д) 2 + 5 % 3
- е) 3 * 4 // 6
- ж) 3 * (4 // 6)
- з) 3 * 2 ** 2
- и) 3 ** 2 * 2


В 2. Все операторы в табл. 1 являются бинарными: они обрабатывают два
операнда (числа) и возвращают одно значение. Символ «минус» (−) также используется как унарный (unary) оператор, который возвращает отрицательное значение (точнее: изменяет знак на противоположный) одного обрабатываемого операнда. Например:

- a = 4
- b = -a
- b

4

Обратите внимание: выражение b = -a (присваивающее переменной b значение a с противоположным знаком) отличается от выражения b -= a (которое вычитает a из b и сохраняет результат в b). Унарный оператор − обладает более высоким приоритетом, чем операторы *, / и %, но более низким приоритетом, чем оператор возведения в степень \**, поэтому, например, -2 ** 4 равно -16
(так как вычисляется −(24), а не (−2)4).
Определить результаты вычислений следующих выражений и проверить их,
используя командную оболочку Python.

- а) -2 ** 2
- б) 2 ** -2
- в) -2 ** -2
- г) 2 ** 2 ** 3
- д)	 2 ** 3 ** 2
- е) -2 ** 3 ** 2
- ж) (-2) ** 3 ** 2
- з) (-2) ** 2 ** 3

**В 3**. Определить и объяснить результаты выполнения следующих инструкций.

- а) 9 + 6j / 2
- б) complex(4, 5).conjugate().imag
- в) complex(0, 3j)
- г) round(2.5)
- д) round(-2.5)
- е) abs(complex(5, -4)) == math.hypot(4,5)

**В 4**. Определить значение $ i^i $ как действительное число, где $ i = \sqrt {−1} $.

**В 5**. Объясните (необычное?) поведение следующего короткого фрагмента
исходного кода:

In [9]:
d = 8
e = 2
from math import *
sqrt(d ** e)

16.88210319127114

**В 6**. Формально операция целочисленного деления a // b определяется как округление в  меньшую сторону (floor) результата деления a/b (иногда такое округление записывают как a/b), т. е. наибольшее целое число, меньшее или
равное a / b. Тогда модуль или остаток от деления a % b (также записываемый как a mod b) равен:

$ a  mod  b = a - b |_a/b_| $

Использовать эти определения для получения результата следующих выражений. Проверить результаты в командной оболочке Python.

- а) 7 // 4
- б) 7 % 4
- в) -7 // 4
- г) -7 % 4
- д)	 7 // -4
- е) 7 % -4
- ж) -7 // -4
- з) -7 % -4


**В 7**. Если две смежные грани правильного шестигранного игрального кубика (кости) имеют значения ```a``` и ```b```, если смотреть сбоку и читать слева направо,то значение на верхней грани кубика вычисляется по формуле 

$ 3( a^3b – ab^3) mod 7 $

Определить значение на верхней грани кубика, если:
- а) a = 2, b = 6;
- б) a = 3, b = 5.

**В 8**. Сколько раз необходимо сложить пополам лист бумаги (толщина t  =  0.1  мм, но можно принять и любой другой размер), чтобы достичь Луны (расстояние от Земли d = 384 400 км)?

**В 9**. Определить результаты вычислений следующих выражений и проверить их, используя командную оболочку Python.

- а) not 1 < 2 or 4 > 2
- б) not (1 < 2 or 4 > 2)
- в) 1 < 2 or 4 > 2
- г) 4 > 2 or 10/0 == 0
- д) not 0 < 1
- е) 1 and 2
- ж)	 0 and 1
- з) 1 or 0
- и) type(complex(2, 3).real) is int


**В 10**. Объясните, почему при вычислении следующего выражения не получается результат 100.

In [13]:
10^2

8

Подсказка: см. документацию Python по битовым операторам (bitwise operators).

# Задачи

**З 1**. В Python нет встроенного оператора «исключающее или», но его можно сформировать из существующих операторов. Разработайте два различных
способа реализации оператора «исключающее или». Таблица истинности для оператора xor приведена в табл. 9.




### Таблица 9. Таблица истинности для оператора xor

|P     |Q      |P xor Q |
|:-----|:------|:-------|
|True  |True   |False   |
|False |True   |True    |
|True  |False  |True    |
|False |False  |False   |

**З 2**. Некоторые интересные вычисления с использованием модуля ```math```:
- а) Какими особенностями обладают числа ```sin(2017 × 5√2)``` и ```(π + 20)i```?
- б) Что произойдет, если попытаться вычислить выражение, такое как ```e1000```,генерирующее число, большее, чем максимальное число с плавающей точкой, которое может быть представлено принятым по умолчанию значением с двойной точностью. Что произойдет, если при вычислении ограничиться целочисленной арифметикой (например, при вычислении 1000!)?
- в) Что произойдет, если попытаться выполнить неопределенную математическую операцию, например деление на нуль?
- г) Максимальное представление числа с  плавающей точкой по стандарту IEEE-754с двойной точностью приблизительно равно 1.8×10308.Вычислить длину гипотенузы прямоугольного треугольника с катетами 1.5×10200 и 3.5×10201:
- i) напрямую используя функцию ```math.hypot()```;
- ii) без использования этой функции.

**З 3**. В некоторых языках есть функция ```sign(a)```, которая возвращает −1, если
аргумент ```a``` отрицательный, и 1, если аргумент ```a``` положительный. В Python такой функции нет, но в  модуль ```math``` включена функция ```math.copysign(x, y)```,которая возвращает абсолютное значение ```x``` со знаком ```y```. Как можно было бы
использовать эту функцию способом, аналогичным использованию отсутствующей функции ```sign(a)```?

**З 4**. Всемирная геодезическая система (сеть) (World Geodetic System) – это комплект международных стандартов для описания формы Земли. В  самой последней версии WGS-84 земной геоид приближенно определяется как эллипсоид, принимающий форму сжатого у  полюсов сфероида с  главной, или большой, полуосью эллипса ```a```  = 6  378  137.0 м  и малой полуосью эллипса ```c``` =
6 356 752.314245 м.

Использовать формулу вычисления площади поверхности сжатого у полюсов сфероида

$ S = 2 pi a^2(1+((1-e^2)/e)atanh(e)) $, где $ e^2 = 1 - (c^2/a^2)  $ 

для вычисления площади поверхности вышеописанного эллипсоида и  сравнить полученный результат с площадью поверхности Земли при предположении, что Земля – сфера с радиусом 6371 км.

